README.md
# The Quiz Board

The aim of this project is the development of a microcontroller based
board to perfom a quiz or puzzle in public. This board is developed
with NGO's in mind, working with the audience on the streets.

The quiz is based on questions with 4 answers. Only one answer is
correct. There are up to 8 questions possible. The sketch below show
a possible design of the board. The buttons could be places within the gray area.

![sketch](https://raw.github.com/joede/quiz-board/master/docs/images/Puzzleboard-Sample--en.png)

The answer is given by pressing one of four buttons. The result is
shown immediately through a red or green LED. The puzzle is done as soon
as all questions got an answer. As an additional option, a time limit
can be enabled, so that all answers must be given within this time period.

To show the state of the operation of the board, a system wide yellow LED is planned.
The sketch above doesn't show this LED.

Since we need a larger number of IOs, we will use a
[Arduino Mega 2560](http://arduino.cc/en/Main/arduinoBoardMega2560)
board.

One important goal of the project is a simple and lean design. It should be easy
for new and unexperienced developers to get familiar with the code. For this
reason, I try to avoid heavy weight libraries in favour of simple and straight
solutions (remember the [KISS](http://en.wikipedia.org/wiki/KISS_principle)
principle).


## Features

Planned features are:

* stand-alone usage without a PC
* selectable set of *correct answers*
* power-safe mode for running on batteries
* optional observation of the time to answer all questions (timeout)

**Note:** the main audience of this project is the German NGO [BUND](http://www.bund.net),
a environmental (protection) organisation. Because of this, all further documents
are written in German. If there is interest from outside of Germany, I may
add an english translation. Feel free to send me your translations.


## License

### Software

EN: Unless otherwise stated, all software or program code and all snippets are licensed under the [GNU GPL 3.0](http://www.gnu.org/licenses/gpl).

DE: Sofern nicht anders angegeben, stehen alle Programme oder Programcode sowie alle Code-Schnipsel unter der [GNU GPL 3.0](http://www.gnu.org/licenses/gpl).

### Documentation

EN: Unless otherwise stated, all documents like manuals, specifications or instructions are licensed under a [Creative Commons Attribution-NonCommercial-ShareAlike 3.0 Unported License](href="http://creativecommons.org/licenses/by-nc-sa/3.0/).

DE: Sofern nicht anderes angegeben, stehen alle Dokumente wie Handbücher, Anleitungen oder Spezifikationen unter einer [Creative Commons Namensnennung - Nicht-kommerziell - Weitergabe unter gleichen Bedingungen 3.0 Unported Lizenz](http://creativecommons.org/licenses/by-nc-sa/3.0/deed.de").

![Creative Commons License](http://i.creativecommons.org/l/by-nc-sa/3.0/88x31.png)


HISTORY.md
# History of 'Quizoard'

## Release 1.0

This is the first release of the project. It includes the regular game play,
optional *result animation* and a optional *game timeout*. The last two
features are not enabled by default.

This releases doesn't implement a *power saving mode* nor a´does is implement
a interactive configuration.


src/config.h
/* QuizBoard -- config.h
 * ========================================================================
 * Copyright 2013 Joerg Desch <github@jdesch.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * ========================================================================
 *
 *
 *
 *
 *
 * ========================================================================
 */
#ifndef __QB__CONFIG_H__
#define __QB__CONFIG_H__

/* Uncomment to build the DEBUGGING release. A debuggin release enables log
 * messages, shortens timeouts and changes the solution set to a simple one.
 */
//#define DEBUG 1


#define APP_VERSION  "1.0"


/*             .-----------------------------------------------.             */
/* ___________/  CONFIGURATION OF THE SOURCE                    \___________ */
/*            `-------------------------------------------------'            */
//{{{

#define CFG_MAX_ANSWERS       4
#define CFG_MAX_QUESTIONS     8
#define CFG_MAX_SOLUTIONSETS  4

#if DEBUG
#  define CFG_GAME_TIMEOUT		(30)		// DEBUG
#  define CFG_GAME_TIMEOUT_WARN		(20)		// DEBUG
#else
#  define CFG_GAME_TIMEOUT		(3*60)		// 3 minutes game play (22s per answer)
#  define CFG_GAME_TIMEOUT_WARN		(2*60)		// warn after 2 minutes
#endif

#define CFG_RESULT_TIMEOUT		(30)		// show result for 1/2 minute


//}}}

/*             .-----------------------------------------------.             */
/* ___________/  CONFIGURATION OF THE MODULES                   \___________ */
/*            `-------------------------------------------------'            */
//{{{

/*  Module "SimpleLogging"
 */
// the maximum length of the string used to build one log message
//#define SZ_QB_SIMPLELOG_BUFFER 50

/*  Module "SimpleTimer"
 */
// the maximum number of concurrently running timers.
#define SZ_QB_MAX_TIMERS 8

//}}}

/*             .-----------------------------------------------.             */
/* ___________/  GENERAL MACRO DEFINITIONS                      \___________ */
/*            `-------------------------------------------------'            */
//{{{

#if defined(__GNUC__)
#  define UNUSED(x) x __attribute__((unused))
#  define ATTR_PACKED __attribute__((__packed__))
#  define ALLWAYS_INLINE inline __attribute__((gnu_inline, always_inline))
#  define PRINTF_STYLE_CXX(idx_fmt,idx_args) __attribute__((format (printf,(idx_fmt)+1,(idx_args)+1)))   // for C++ methods onyl
#  define PRINTF_STYLE(idx_fmt,idx_args) __attribute__((format (printf,idx_fmt,idx_args)))               // for regular functions
#else
#  define UNUSED(x) x
#  define ATTR_PACKED
#  define PRINTF_STYLE(idx_fmt,idx_args)
#  define ALWAYS_INLINE
#endif

//}}}

/*             .-----------------------------------------------.             */
/* ___________/  GENERIC STUFF                                  \___________ */
/*            `-------------------------------------------------'            */
//{{{
//}}}

#endif // __QB__CONFIG_H__


src/QuizGame.h
/* QuizBoard - QuizGame.h
 * ========================================================================
 * Copyright 2013 Joerg Desch <github@jdesch.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * ========================================================================
 *
 * This class implements logic of the quiz game. It handles the different
 * answer sets, checks the answers and keep track of all answered questions.
 *
 * Note: the API of his class counts all answers and questions from 1.
 *
 * ========================================================================
 * TODO:
 *
 * ========================================================================
 */
#ifndef __QB_QUIZGAME_HXX__
#define __QB_QUIZGAME_HXX__

/* These local include files must be opened in a tab. If not, the headers
 * could not be found! Don't aks me why...
 */
#include "config.h"

#include <stdarg.h>
#include <Arduino.h>

#include "SimpleLogging.h"

#ifndef CFG_MAX_ANSWERS
#  error "fatal: CFG_MAX_ANSWERS must be defined in config.h"
#endif
#ifndef CFG_MAX_QUESTIONS
#  error "fatal: CFG_MAX_QUESTIONS must be defined in config.h"
#endif
#ifndef CFG_MAX_SOLUTIONSETS
#  error "fatal: CFG_MAX_SOLUTIONSETS must be defined in config.h"
#endif


/* The "Quiz Game " class
 *
 */
class QuizGame
{
  /* --------------------------------------------
   * inner data types
   */
  private:
    struct SolutionSet { int CorrectAnswer[CFG_MAX_QUESTIONS]; };

  public:
    enum RESULTS {RES_ERROR=(-1), RES_NONE=0, RES_WRONG, RES_CORRECT };

  /* --------------------------------------------
   * constants declarations
   */
  //private:
  //public:

  /* --------------------------------------------
   * local attributes
   */
  protected:
    static SolutionSet m_SolutionSet[CFG_MAX_SOLUTIONSETS];
    int m_GivenAnswer[CFG_MAX_QUESTIONS];	// the given answers. 0=unanswered
    int m_NumUnAnswered;			// number of questions without an answer
    int m_UseSet;				// used answer set
    bool m_GameRunning;
    SimpleLogging m_dbg;

/* --------------------------------------------
   * methods
   */
  protected:


  public:

    QuizGame();
    void stopGame ( void );
    void startGame ( int AnswerSet );
    bool gameFinished ( void );
    bool giveAnswer ( int question, int answer );
    int questionAnswered ( int question );
    bool success ( void );
    bool gameRunning (void) {return m_GameRunning;};
};


/* If this define is defined, we can allocate our code and static data here.
 * This way we avoid libraries and additional CXX files.
 */
#ifdef __ALLOC_STATICS_HERE__

QuizGame::QuizGame()
{
#ifdef DEBUG
    m_dbg.setLevel(SimpleLogging::LVL_DEBUG);
#else
    m_dbg.setLevel(SimpleLogging::LVL_WARNING);
#endif
    stopGame();
}

void QuizGame::stopGame ( void )
{
    m_GameRunning = false;
    m_UseSet = 0;
    m_NumUnAnswered = CFG_MAX_QUESTIONS;
    for ( int i=0; i<CFG_MAX_QUESTIONS; i++ )
	m_GivenAnswer[i] = 0;
}

void QuizGame::startGame ( int AnswerSet )
{
    m_dbg.log(SimpleLogging::LVL_DEBUG,"QuizGame::startGame");
    if ( AnswerSet<0 || AnswerSet>=CFG_MAX_SOLUTIONSETS )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"QuizGame::startGame: invalid AnswerSet! use default.");
	m_UseSet = 0;
    }
    else
        m_UseSet = AnswerSet;
    m_NumUnAnswered = CFG_MAX_QUESTIONS;
    for ( int i=0; i<CFG_MAX_QUESTIONS; i++ )
	m_GivenAnswer[i] = 0;
    m_GameRunning = true;
}

/* Returns true is the game is finished. The game is finished, if there
 * are no more unanswered questions left.
 */
bool QuizGame::gameFinished ( void )
{
    return (m_NumUnAnswered == 0) ? true:false;
}

/* Pass a given answer to the game. The method returns false, if the answer
 * is not excepted. After counting a valid answer, the answer is stored
 * Use questionAnswered() to get the result. Use gameFinished() to check if the
 * the game is done.
 */
bool QuizGame::giveAnswer ( int question, int answer )
{
    if ( !m_GameRunning )
    {
	m_dbg.log(SimpleLogging::LVL_WARNING,"QuizGame::giveAnswer: Game not running.");
	return false;
    }
    if ( question<=0 || question>CFG_MAX_QUESTIONS )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"QuizGame::giveAnswer: invalid question.");
	return false;
    }
    if ( answer<=0 || answer>CFG_MAX_ANSWERS )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"QuizGame::giveAnswer: invalid answer.");
	return false;
    }
    if ( m_NumUnAnswered==0 )
    {
	m_dbg.log(SimpleLogging::LVL_DEBUG,"QuizGame::giveAnswer: nothing to do");
	return true;
    }

    // we need an index which count from zero.
    --question;

    if ( m_GivenAnswer[question]==0 )		// store only if not yet answered
    {
	m_GivenAnswer[question] = answer;	// the the answer
	--m_NumUnAnswered;			// and count it.
    }
    else
    {
	m_dbg.log(SimpleLogging::LVL_DEBUG,"QuizGame::giveAnswer: question already answered");
    }
    return true;
}

/* Check if the given question is answered and if the answer is correct.
 * The returned value is a RES_* code.
 *
 *   RES_ERROR    =>  (-1) error. invalid parameter
 *   RES_NONE     =>  (0)  question isn't answered yet
 *   RES_WRONG    =>  (1)  wrong answer
 *   RES_CORRECT  =>  (2)  correct answer
 */
int QuizGame::questionAnswered ( int question )
{
    int CorrectAnswer;

    if ( question<=0 || question>CFG_MAX_QUESTIONS )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"QuizGame::questionAnswered: invalid parameter.");
	return RES_ERROR;
    }
    // we need an index which count from zero.
    --question;

    // check if answered yet
    if ( m_GivenAnswer[question]==0 )
	return RES_NONE;

    // compare the result
    CorrectAnswer = m_SolutionSet[m_UseSet].CorrectAnswer[question];
    // m_dbg.log(SimpleLogging::LVL_DEBUG,"QuizGame::questionAnswered: %d==%d?",m_GivenAnswer[question],CorrectAnswer);
    return (m_GivenAnswer[question]==CorrectAnswer) ? RES_CORRECT:RES_WRONG;
}

/* The game is won, if all questions are answered. Simple, isn't it. ;-)
 */
bool QuizGame::success ( void )
{
    for ( int i=1; i<=CFG_MAX_QUESTIONS; i++ )
    {
	if ( questionAnswered(i)!=RES_CORRECT )
	    return false;
    }
    return true;
}


QuizGame::SolutionSet QuizGame::m_SolutionSet[CFG_MAX_SOLUTIONSETS] =
{
#ifdef DEBUG
    // these sets are just debugging sets...
    {1,2,3,4,1,2,3,4},
    {4,3,2,1,4,3,2,1},
    {1,2,1,2,3,4,3,4},
    {4,3,4,3,2,1,2,1}
#else
    // set #0
    {4,1,4,3,2,1,4,1},
    // set #1
    {4,3,3,1,1,3,1,2},
    // set #2
    {1,3,2,2,1,2,4,3},
    // set #3
    {1,3,1,1,4,4,1,3}
#endif
};

#endif // __ALLOC_STATICS_HERE__
#endif // __QB_QUIZGAME_HXX__


tools/generate-sets.c
//
//
// COMPILE:  gcc -Wall -std=c99 -o generate-sets generate-sets.c
//
// USAGE:    ./generate-sets 14 4 8 2 c

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

int main ( int argc, char *argv[] )
{
    int answers;
    int questions;
    int sets;
    unsigned int seed;
    bool c_source;
    int r;

    if ( argc != 6 )
    {
        fprintf(stderr, "Usage: %s <seed> <answers> <questions> <sets> {c|t}\n", argv[0]);
        exit(EXIT_FAILURE);
    }
    if ( argv[5][0] != 'c' && argv[5][0] != 't' )
    {
        fprintf(stderr, "Usage: %s <seed> <answers> <questions> <sets> {c|t}\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    seed = atoi(argv[1]);
    answers = atoi(argv[2]);
    questions = atoi(argv[3]);
    sets = atoi(argv[4]);
    c_source = ( argv[5][0] == 'c' ) ? true:false;
    if ( seed==0 || answers==0 || questions==0 || sets==0 )
    {
        fprintf(stderr, "error: invalid numerical parameter\n");
        exit(EXIT_FAILURE);
    }

    if ( c_source )
	printf("    {\n");
    srand(seed);
    for ( int s=0; s<sets; s++)
    {
	if ( c_source )
	    printf("        // set #%d\n        { ",s);
	for ( int q=0; q<questions; q++)
	{
	    r =  (int)rand() % answers + 1;
	    // printf("%d ==> mod4: %d   div: %d\n", r, r%4+1, (int)(((double)r/(double)RAND_MAX)*4.0+0.5) );
	    if ( c_source )
		printf("%d%s",r,(q==(questions-1))?"":",");
	    else
		printf("%d%s",r,(q==(questions-1))?"":";");
	}
	if ( c_source )
	    printf("}%s\n",(s==(sets-1))?"":",");
	else
	    printf("\n");
    }
    if ( c_source )
	printf("    }\n");
    exit(EXIT_SUCCESS);
}


src/QuizBoard.h
/* QuizBoard -- QuizBoard.h
 * ========================================================================
 * Copyright 2013 Joerg Desch <github@jdesch.de>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA 02110-1301, USA.
 * ========================================================================
 *
 * The class QuizBoard describes the hardware of the project. In this case
 * this means the keypad and the LEDs. The class distiguishs between regular
 * keys and answer keys. The differtiation is made by an specially formed
 * *key code*. This code is above QuizBoard::OffsetRegularKeys for regular
 * keys. For answer keys, the code is below this value. In addition to that,
 * the code of the answer keys is built by the number of the answer and
 * the number of the question the answer belongs to.
 *
 * ========================================================================
 */
#ifndef __QB_QUIZBOARD_HXX__
#define __QB_QUIZBOARD_HXX__


/* These local include files must be opened in a tab. If not, the headers
 * could not be found! Don't aks me why...
 */
#include "config.h"
#include <Arduino.h>

#include "SimpleLogging.h"

#ifndef CFG_MAX_ANSWERS
#  error "fatal: CFG_MAX_ANSWERS must be defined in config.h"
#endif
#if CFG_MAX_ANSWERS > 4
#  error "todo: change code to allow CFG_MAX_ANSWERS larger than 4"
#endif

#ifndef CFG_MAX_QUESTIONS
#  error "fatal: CFG_MAX_QUESTIONS must be defined in config.h"
#endif


/*
 *
 */
class QuizBoard
{
  /* --------------------------------------------
   * inner data types
   */
  public:
    enum LED {LED_RED1=0, LED_RED2, LED_RED3, LED_RED4, LED_RED5, LED_RED6, LED_RED7, LED_RED8,
              LED_GREEN1, LED_GREEN2, LED_GREEN3, LED_GREEN4, LED_GREEN5, LED_GREEN6, LED_GREEN7, LED_GREEN8,
              LED_STATUS, LED_ALIVE };
    enum ANSWER_STATES {ANS_NONE=0, ANS_WRONG, ANS_CORRECT, MAX_ANSWER_STATES};
    enum KEY {KEY_REGULAR_KEYS=100, KEY_START };
    struct KeyDefinition { int pin; int keycode; };


  /* --------------------------------------------
   * constants declarations
   */
  private:
    const int DebounceTime = 50;                     // [ms] for debouncing keys
    const int OffsetRegularKeys = KEY_REGULAR_KEYS;  // key code were regular keys starts

  public:
    static const int KEY_NONE = -1;
    static const int MAX_ANSWERS = CFG_MAX_ANSWERS;
    static const int MAX_QUESTIONS = CFG_MAX_QUESTIONS;


  /* --------------------------------------------
   * local attributes
   */
  protected:
    int NumOfKeys;
    static KeyDefinition ListOfKeys[];
    int NumOfLEDs;
    static int ListOfLEDs[];
    int CurrentKeyPressed;
    int CurrentKeyChecked;
    long LastReadTime;
    SimpleLogging m_dbg;



  /* --------------------------------------------
   * methods
   */
  public:

    QuizBoard()
    {
	NumOfKeys = 0;                            // not detected
	NumOfLEDs = 0;                            // not detected
	CurrentKeyPressed = KEY_NONE;             // no key pressed
	CurrentKeyChecked = KEY_NONE;             // no key checked
	LastReadTime = 0;
    };

    bool dbg_Check(void)
    {
        if ( CurrentKeyPressed!=KEY_NONE && CurrentKeyChecked!=KEY_NONE )
        {
            m_dbg.log(SimpleLogging::LVL_CRITICAL,"illegal transit state!");
        }
    };

    // to which question belongs the key (0=none)
    int isQuestion ( int KeyNr )
    {
        int i;
        if ( KeyNr < 0 || KeyNr >= OffsetRegularKeys )
	    return 0;
        i = (KeyNr>>2)+1;
        m_dbg.log(SimpleLogging::LVL_DEBUG,"k=%d q=%d",KeyNr,i);
        return i>MAX_QUESTIONS ? 0 : i;
    };

    // to which answer of the question belongs the key (0=none)
    int isAnswer ( int KeyNr )
    {
	int i;
	if ( KeyNr < 0 || KeyNr >= OffsetRegularKeys )
	    return 0;
	i = (KeyNr&0x0003)+1;
	m_dbg.log(SimpleLogging::LVL_DEBUG,"k=%d a=%d",KeyNr,i);
	return i>MAX_ANSWERS ? 0 : i;
    };

    bool isRegularKey ( int KeyNr )
    {
	return ( KeyNr >= OffsetRegularKeys ) ? true : false;
    }

    bool setup ( void );

    void disableLEDs ( void );

    void setLED ( int nr, bool new_state );

    void showAnswer ( int question, int AnswerState );

    int waitKey ( void );
};


/* If this define is defined, we can allocate our code and static data here.
 * This way we avoid libraries and additional CXX files.
 */
#ifdef __ALLOC_STATICS_HERE__

bool QuizBoard::setup ( void )
{
    int i;

#ifdef DEBUG
    m_dbg.setLevel(SimpleLogging::LVL_DEBUG);
#else
    m_dbg.setLevel(SimpleLogging::LVL_WARNING);
#endif
    m_dbg.log(SimpleLogging::LVL_DEBUG,"QuizBoard::setup");

    /* count the number of keys.
     */
    for ( i=0; ListOfKeys[i].pin>=0; i++ )
    {
	if ( i>=60 )
	{
	    i = 0;
	    break;
	}
    }
    if ( i <= 0 )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"ERROR: QuizBoard: illegal number of keys in list!");
	return false;
    }
    else
	NumOfKeys = i;
    m_dbg.log(SimpleLogging::LVL_DEBUG,"Found %d key definitions!",NumOfKeys);

    /* count the number of LEDs.
     */
    for ( i=0; ListOfLEDs[i]>=0; i++ )
    {
	if ( i>=60 )
	{
	    i = 0;
	    break;
	}
    }
    if ( i <= 0 )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"ERROR: QuizBoard: illegal number of LEDs in list!");
	return false;
    }
    else
	NumOfLEDs = i;
    m_dbg.log(SimpleLogging::LVL_DEBUG,"Found %d LED definitions!",NumOfLEDs);

    // loop through the list and set INPUTs nad OUTPUTs
    for ( i=0; i<NumOfKeys; i++ )
    {
	pinMode(ListOfKeys[i].pin,INPUT_PULLUP);  // input with internal pullup enabled
    }
    for ( i=0; i<NumOfLEDs; i++ )
    {
	pinMode(ListOfLEDs[i],OUTPUT);            // output for LEDs; low means on
	digitalWrite(ListOfLEDs[i],HIGH);         // disable LED
    }

    return true;
}

void QuizBoard::disableLEDs ( void )
{
    for ( int i=0; i<NumOfLEDs; i++ )
    {
	digitalWrite(ListOfLEDs[i],HIGH);         // disable LED
    }
}

/* Simple set one of the LEDs.
 *
 */
void QuizBoard::setLED ( int nr, bool new_state )
{
    if ( nr>=0 && nr<=NumOfLEDs )
    {
	digitalWrite(ListOfLEDs[nr],new_state?LOW:HIGH);
    }
    else
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"ERROR: QuizBoard::setLED: bad parm\n");
    }
}

/* Set the LEDs of the passed question to the passed state.
 */
void QuizBoard::showAnswer ( int question, int AnswerState )
{
    if ( AnswerState<ANS_NONE || AnswerState>=MAX_ANSWER_STATES )
    {
	m_dbg.log(SimpleLogging::LVL_ERROR,"ERROR: QuizBoard::showAnswer: bad parm\n");
	return;
    }
    switch (question)
    {
	case 1:
	    setLED(LED_RED1,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN1,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 2:
	    setLED(LED_RED2,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN2,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 3:
	    setLED(LED_RED3,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN3,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 4:
	    setLED(LED_RED4,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN4,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 5:
	    setLED(LED_RED5,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN5,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 6:
	    setLED(LED_RED6,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN6,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 7:
	    setLED(LED_RED7,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN7,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	case 8:
	    setLED(LED_RED8,(AnswerState==ANS_WRONG)?true:false);
	    setLED(LED_GREEN8,(AnswerState==ANS_CORRECT)?true:false);
	    break;
	default:
	    m_dbg.log(SimpleLogging::LVL_ERROR,"ERROR: QuizBoard::showAnswer: bad parm\n");
    }
}


/* Wait for a key to be pressed. The newly pressed key is returned. If no key
 * is pressed or if the previous key is hold pressed, the function returns
 * KEY_NONE (-1).
 *
 * Restrictions: only one key can be pressed at the same time! The priority is
 * ------------- the order inside the list ListOfKeys[].
 */
int QuizBoard::waitKey (void)
{
    int kstate;
    int i;

    // is the previous key still pressed? Wait for it's release
    if ( CurrentKeyPressed != KEY_NONE )
    {
        kstate = digitalRead(ListOfKeys[CurrentKeyPressed].pin);
        //m_dbg.log(SimpleLogging::LVL_DEBUG,"**rel?=%d",CurrentKeyPressed);
        if ( kstate == HIGH )
        {
            //m_dbg.log(SimpleLogging::LVL_DEBUG,"**release=#%d",CurrentKeyPressed);
	    CurrentKeyPressed = KEY_NONE;
            CurrentKeyChecked = KEY_NONE;
        }
        return KEY_NONE;                             // still pressed, so there's no change.
    }
    else
    {
        // is no key pressed?
        if ( CurrentKeyChecked == KEY_NONE )
        {
            // Look for a pressed one.
            for ( i=0; ListOfKeys[i].pin>=0; i++ )
            {
                kstate = digitalRead(ListOfKeys[i].pin);
                LastReadTime = millis();
                if ( kstate == LOW )
                {
                    // stop searching, we have found a candidate.
                    //m_dbg.log(SimpleLogging::LVL_DEBUG,"**new?=%d",i);
                    CurrentKeyChecked = i;
                    return KEY_NONE;                       // still not "officially" pressed..
                }
            }
        }
        else
        {
            // or check the debounce for this key.
            kstate = digitalRead(ListOfKeys[CurrentKeyChecked].pin);
            if ( kstate==LOW && ((millis()-LastReadTime) > DebounceTime) )
            {
                CurrentKeyPressed = CurrentKeyChecked;
                CurrentKeyChecked = KEY_NONE;
            }
            else if ( kstate==HIGH )
            {
                // the key isn't pressed anymore
                CurrentKeyChecked = KEY_NONE;
                //m_dbg.log(SimpleLogging::LVL_DEBUG,"**lost=%d",i);
            }
        }
    }
    //if ( CurrentKeyPressed != KEY_NONE )
    //    m_dbg.log(SimpleLogging::LVL_DEBUG,"**pressed=#%d",CurrentKeyPressed);
    return ListOfKeys[CurrentKeyPressed].keycode;
}

/* generate a keycode out of the question and answer numbers.
 * Both parameters are counting from 1 but stored counting from 0!
 */
#define QB_GENKCODE(question,answer) (((((question)-1)&0x0007)<<2)|(((answer)-1)&0x0003))

/* List of pin numbers with all the keys. The order must match the key
 * code of the enum QuizBoard::KEY!
 */
QuizBoard::KeyDefinition QuizBoard::ListOfKeys[] =
{
    // answer keys for question  #8
    { pin:22, keycode:QB_GENKCODE(8,4) },
    { pin:23, keycode:QB_GENKCODE(8,3) },
    { pin:24, keycode:QB_GENKCODE(8,2) },
    { pin:25, keycode:QB_GENKCODE(8,1) },

    // answer keys for question  #7
    { pin:26, keycode:QB_GENKCODE(7,4) },
    { pin:27, keycode:QB_GENKCODE(7,3) },
    { pin:28, keycode:QB_GENKCODE(7,2) },
    { pin:29, keycode:QB_GENKCODE(7,1) },

    // answer keys for question  #6
    { pin:30, keycode:QB_GENKCODE(6,4) },
    { pin:31, keycode:QB_GENKCODE(6,3) },
    { pin:32, keycode:QB_GENKCODE(6,2) },
    { pin:33, keycode:QB_GENKCODE(6,1) },

    // answer keys for question  #5
    { pin:34, keycode:QB_GENKCODE(5,4) },
    { pin:35, keycode:QB_GENKCODE(5,3) },
    { pin:36, keycode:QB_GENKCODE(5,2) },
    { pin:37, keycode:QB_GENKCODE(5,1) },

    // answer keys for question  #4
    { pin:38, keycode:QB_GENKCODE(4,4) },
    { pin:39, keycode:QB_GENKCODE(4,3) },
    { pin:40, keycode:QB_GENKCODE(4,2) },
    { pin:41, keycode:QB_GENKCODE(4,1) },

    // answer keys for question  #3
    { pin:42, keycode:QB_GENKCODE(3,4) },
    { pin:43, keycode:QB_GENKCODE(3,3) },
    { pin:44, keycode:QB_GENKCODE(3,2) },
    { pin:45, keycode:QB_GENKCODE(3,1) },

    // answer keys for question  #2
    { pin:46, keycode:QB_GENKCODE(2,4) },
    { pin:47, keycode:QB_GENKCODE(2,3) },
    { pin:48, keycode:QB_GENKCODE(2,2) },
    { pin:49, keycode:QB_GENKCODE(2,1) },

    // answer keys for question  #1
    { pin:50, keycode:QB_GENKCODE(1,4) },
    { pin:51, keycode:QB_GENKCODE(1,3) },
    { pin:52, keycode:QB_GENKCODE(1,2) },
    { pin:53, keycode:QB_GENKCODE(1,1) },

    // other (regular) keys
    { pin:8, keycode:KEY_START },

    { pin:-1, keycode:-1 }
};


/* List of pin numbers with all the LEDs. The order must match the LED
 * number of the enum QuizBoard::LED!
 */
int QuizBoard::ListOfLEDs [] =
{
    // "red" LEDs for wrong answers
    A0, A1, A2, A3, A4, A5, A6, A7,
    // "green" LEDs for correct answers
    A8, A9, A10, A11, A12, A13, A14, A15,
    // other LEDs
    11, 13,					// LED_STATUS, LED_ALIVE
    -1
};

#endif // __ALLOC_STATICS_HERE__
#endif // __QUIZBOARD_HXX__
